### PR TITLE
JIT: Optimize SequenceEqual to use ccmp on ARM64

### PR DESCRIPTION
**Pull request by jakobbotsch**
In the original PR we could not get this this working due to some conservative interference. This now does the right thing with #92710 merged.

Also change LowerCallMemcmp/LowerCallMemmove to return next node to lower just to align it a bit more with other functions.

```csharp
bool Test1(ReadOnlySpan<byte> data) => "hello world"u8.SequenceEqual(data);
```

Before:
```asm
G_M52062_IG01:  ;; offset=0x0000
            stp     fp, lr, [sp, #-0x10]!
            mov     fp, sp
						;; size=8 bbWeight=1 PerfScore 1.50
G_M52062_IG02:  ;; offset=0x0008
            movz    x2, #0x20D4      // data for <PrivateImplementationDetails>:430646847E70344C09F58739E99D5BC96EAC8D5FE7295CF196B986279876BF9B
            movk    x2, #0xE179 LSL #16
            movk    x2, #554 LSL #32
            cmp     w1, #11
            bne     G_M52062_IG05
						;; size=20 bbWeight=1 PerfScore 3.00
G_M52062_IG03:  ;; offset=0x001C
            ldr     x1, [x2]
            ldr     x2, [x2, #0x03]
            ldr     x3, [x0]
            ldr     x0, [x0, #0x03]
            eor     x1, x1, x3
            eor     x0, x2, x0
            orr     x0, x1, x0
            cmp     x0, #0
            cset    x0, eq
						;; size=36 bbWeight=0.48 PerfScore 6.96
G_M52062_IG04:  ;; offset=0x0040
            ldp     fp, lr, [sp], #0x10
            ret     lr
						;; size=8 bbWeight=1 PerfScore 2.00
G_M52062_IG05:  ;; offset=0x0048
            mov     w0, wzr
            b       G_M52062_IG04
						;; size=8 bbWeight=0 PerfScore 0.00
```

After:
```asm
G_M52062_IG01:  ;; offset=0x0000
            stp     fp, lr, [sp, #-0x10]!
            mov     fp, sp
						;; size=8 bbWeight=1 PerfScore 1.50
G_M52062_IG02:  ;; offset=0x0008
            movz    x2, #0x20D4      // data for <PrivateImplementationDetails>:430646847E70344C09F58739E99D5BC96EAC8D5FE7295CF196B986279876BF9B
            movk    x2, #0x63E LSL #16
            movk    x2, #445 LSL #32
            cmp     w1, #11
            bne     G_M52062_IG05
						;; size=20 bbWeight=1 PerfScore 3.00
G_M52062_IG03:  ;; offset=0x001C
            ldr     x1, [x2]
            ldr     x2, [x2, #0x03]
            ldr     x3, [x0]
            ldr     x0, [x0, #0x03]
            cmp     x1, x3
            ccmp    x2, x0, 0, eq
            cset    x0, eq
						;; size=28 bbWeight=0.48 PerfScore 6.48
G_M52062_IG04:  ;; offset=0x0038
            ldp     fp, lr, [sp], #0x10
            ret     lr
						;; size=8 bbWeight=1 PerfScore 2.00
G_M52062_IG05:  ;; offset=0x0040
            mov     w0, wzr
            b       G_M52062_IG04
						;; size=8 bbWeight=0 PerfScore 0.00
```

Benchmarks from #83945 on my M1:
|       Method |        Job |               Toolchain |      Mean |     Error |    StdDev | Ratio |
|------------- |----------- |------------------------ |----------:|----------:|----------:|------:|
|      TE_Json | Job-RFUZAM |      /Core_Root/corerun | 1.9945 ns | 0.0100 ns | 0.0083 ns |  0.98 |
|      TE_Json | Job-KSXFQI | /Core_Root_base/corerun | 2.0339 ns | 0.0011 ns | 0.0009 ns |  1.00 |
|              |            |                         |           |           |           |       |
| TE_Plaintext | Job-RFUZAM |      /Core_Root/corerun | 0.7460 ns | 0.0086 ns | 0.0080 ns |  0.99 |
| TE_Plaintext | Job-KSXFQI | /Core_Root_base/corerun | 0.7515 ns | 0.0008 ns | 0.0008 ns |  1.00 |
|              |            |                         |           |           |           |       |
|    Equals_15 | Job-RFUZAM |      /Core_Root/corerun | 0.5063 ns | 0.0008 ns | 0.0008 ns |  0.87 |
|    Equals_15 | Job-KSXFQI | /Core_Root_base/corerun | 0.5799 ns | 0.0006 ns | 0.0006 ns |  1.00 |

### CODE REVIEW

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @EgorBo 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=422680&view=results)

File: src/coreclr/jit/lower.cpp [hunk 1]
```diff
@@ -2020,51 +2023,77 @@ GenTree* Lowering::LowerCallMemcmp(GenTreeCall* call)
...
 
                     BlockRange().InsertAfter(rArgClone, l1Indir, l2Offs, l2AddOffs, l2Indir);
                     BlockRange().InsertAfter(l2Indir, r1Indir, r2Offs, r2AddOffs, r2Indir);
-                    BlockRange().InsertAfter(r2Indir, lXor, rXor, resultOr, zeroCns);
-                    BlockRange().InsertAfter(zeroCns, result);
+
+#ifdef TARGET_ARM64
+                    if (!varTypeIsSIMD(loadType))
+                    {
+                        // ARM64 will get efficient ccmp codegen if we emit the normal thing:
```
Comment (EgorBo) [inline src/coreclr/jit/lower.cpp:2043]: I remember we discussed this and it didn't want to emit ccmp due to some side effects or something like that (probably my PR that re-ordered INDirs fixed that?)

File: src/coreclr/jit/lower.cpp [hunk 2]
```diff
@@ -2020,51 +2023,77 @@ GenTree* Lowering::LowerCallMemcmp(GenTreeCall* call)
...
+                        //
+                        // bool result = (*(int*)leftArg == *(int)rightArg) & (*(int*)(leftArg + 1) == *(int*)(rightArg
+                        // +
+                        // 1))
+
+                        GenTree* eq1 = newBinaryOp(comp, GT_EQ, TYP_INT, l1Indir, r1Indir);
+                        GenTree* eq2 = newBinaryOp(comp, GT_EQ, TYP_INT, l2Indir, r2Indir);
+                        result       = newBinaryOp(comp, GT_AND, TYP_INT, eq1, eq2);
+
+                        BlockRange().InsertAfter(r2Indir, eq1, eq2, result);
```
Comment (EgorBo) [inline src/coreclr/jit/lower.cpp:2053]: Ideally, it should work for SIMD as well and then be optimized: https://github.com/dotnet/runtime/issues/89722

File: src/coreclr/jit/lower.cpp [hunk 3]
```diff
@@ -2020,51 +2023,77 @@ GenTree* Lowering::LowerCallMemcmp(GenTreeCall* call)
...
+                    *next = lArgClone;
+
                     GenTree* l1Indir   = comp->gtNewIndir(loadType, lArgUse.Def());
                     GenTree* r1Indir   = comp->gtNewIndir(loadType, rArgUse.Def());
-                    GenTree* lXor      = newBinaryOp(comp, GT_XOR, actualLoadType, l1Indir, r1Indir);
                     GenTree* l2Offs    = comp->gtNewIconNode(cnsSize - loadWidth, TYP_I_IMPL);
                     GenTree* l2AddOffs = newBinaryOp(comp, GT_ADD, lArg->TypeGet(), lArgClone, l2Offs);
                     GenTree* l2Indir   = comp->gtNewIndir(loadType, l2AddOffs);
-                    GenTree* r2Offs    = comp->gtCloneExpr(l2Offs); // offset is the same
+                    GenTree* r2Offs    = comp->gtNewIconNode(cnsSize - loadWidth, TYP_I_IMPL);
```
Comment (EgorBo) [inline src/coreclr/jit/lower.cpp:2033]: motivation? 🙂

**Review by EgorBo — APPROVED** [review]

File: src/coreclr/jit/lower.cpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/lower.cpp:2053]: I left a TODO-CQ about it below, presumably both the int/SIMD paths can benefit on xarch (while only SIMD on arm64).
I was thinking we could model it after/combine it with `TryLowerAndOrToCCMP`, although I can see your prototype did it on tree form which probably makes it simpler

File: src/coreclr/jit/lower.cpp (hunk 3 above)
Comment (jakobbotsch) [reply src/coreclr/jit/lower.cpp:2033]: No strong motivation... just that we were already creating all the other constants from scratch so it looked a bit out of place to me to clone one lone one. I can revert it if you prefer

Comment (EgorBo) [reply src/coreclr/jit/lower.cpp:2033]: I am fine with this, was just wondering if you were improving TP impact or something like that

**Pull request merged**